Python/ceval.c

```diff
@@ -4358,13 +4351,98 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
                 }
                 _PyErr_Clear(tstate);
             }
+        iterator_exhausted_no_error:
             /* iterator ended normally */
-            STACK_SHRINK(1);
-            Py_DECREF(iter);
-            JUMPBY(oparg);
+            assert(!_PyErr_Occurred(tstate));
+            Py_DECREF(POP());
+            JUMPBY(INLINE_CACHE_ENTRIES_FOR_ITER + oparg);
             DISPATCH();
         }
 
+        TARGET(FOR_ITER_ADAPTIVE) {
+            assert(cframe.use_tracing == 0);
+            _PyForIterCache *cache = (_PyForIterCache *)next_instr;
+            if (ADAPTIVE_COUNTER_IS_ZERO(cache)) {
+                next_instr--;
+                _Py_Specialize_ForIter(TOP(), next_instr);
+                NOTRACE_DISPATCH_SAME_OPARG();
+            }
+            else {
+                STAT_INC(FOR_ITER, deferred);
+                DECREMENT_ADAPTIVE_COUNTER(cache);
+                JUMP_TO_INSTRUCTION(FOR_ITER);
+            }
+        }
+
+        TARGET(FOR_ITER_LIST) {
+            assert(cframe.use_tracing == 0);
+            _PyListIterObject *it = (_PyListIterObject *)TOP();
+            DEOPT_IF(Py_TYPE(it) != &PyListIter_Type, FOR_ITER);
+            STAT_INC(FOR_ITER, hit);
+            PyListObject *seq = it->it_seq;
+            if (seq == NULL) {
+                goto iterator_exhausted_no_error;
+            }
+            if (it->it_index < PyList_GET_SIZE(seq)) {
+                PyObject *next = PyList_GET_ITEM(seq, it->it_index++);
+                Py_INCREF(next);
+                PUSH(next);
+                JUMPBY(INLINE_CACHE_ENTRIES_FOR_ITER);
+                NOTRACE_DISPATCH();
+            }
+            it->it_seq = NULL;
+            Py_DECREF(seq);
+            goto iterator_exhausted_no_error;
+        }
+
+        TARGET(FOR_ITER_RANGE) {
+            assert(cframe.use_tracing == 0);
+            _PyRangeIterObject *r = (_PyRangeIterObject *)TOP();
+            DEOPT_IF(Py_TYPE(r) != &PyRangeIter_Type, FOR_ITER);
+            STAT_INC(FOR_ITER, hit);
+            _Py_CODEUNIT next = next_instr[INLINE_CACHE_ENTRIES_FOR_ITER];
+            assert(_PyOpcode_Deopt[_Py_OPCODE(next)] == STORE_FAST);
+            PyObject **local_ptr = &GETLOCAL(_Py_OPARG(next));
+            PyObject *local = *local_ptr;
+            if (r->index >= r->len) {
+                goto iterator_exhausted_no_error;
+            }
+            JUMPBY(INLINE_CACHE_ENTRIES_FOR_ITER + 1);
+            sdigit value = r->start + (digit)(r->index++) * r->step;
+            if (value < _PY_NSMALLPOSINTS && value >= -_PY_NSMALLNEGINTS) {
```

Comment:
I think there are too many branches here.
Ideally a specialized instruction should have a few `DEOPT_IF` tests, then the rest of the code should be (mostly) branchless.

This also includes far too much knowledge of the internals of the `int` object, IMO.

Would performance be measurably worse if `start`, `end` and `stop` were `Py_ssize_t` and this code were simplified to 
`PyLong_FromSsize_t`?
This feels like an elaborate workaround the oddities of our `int` implementation.
We should be fixing that, not working around it.

Reply:
I think there may be a middle ground here: add a function like this to longobject.c:

```
int
_PyLong_AssignValueLong(PyLongObject **location, long value)
{
    if value in small:
        set to small, decref old long
    elif refcnt(*location) == 1 and the right size:
        mutate in place
    else:
        Py_SETREF(*location, PyLong_FromLong(value));
}
```

Sort of similar in spirit to `PyUnicode_Append()`

That way, the PyLongObject implementation details stay out of the eval loop, but we still get the benefit of avoiding the allocation, which I believe is the main source of the speedup.

Reply:
This could theoretically be re-used for `enumerate()` as well.

And even if the implementation of integers changes (e.g. new freelists, no more small ints), the operation still makes sense.